test/unit/puppetlabs/puppetserver/lockable_pool_test.clj

```diff
@@ -60,6 +85,27 @@
         ;; make sure we got here
         (is (true? true))))))
 
+(deftest pool-lock-is-blocking-test-until-borrows-unregistered-test
+  (let [pool (create-populated-pool 3)
+        instances (borrow-n-instances pool 2)
+        future-started? (promise)]
+    (is (= 2 (count instances)))
+    (let [lock-thread (future (deliver future-started? true)
+                              (.lock pool))]
+      @future-started?
+      (testing "pool.lock() blocks until all instances are unregistered"
+        (is (not (realized? lock-thread)))
+
+        (testing (str "other threads may successfully unregister instances "
+                      "while pool.lock() is being executed")
+          (.unregister pool (first instances))
+          (is (not (realized? lock-thread)))
+          (.unregister pool (second instances)))
+
+        @lock-thread
+        ;; make sure we got here
+        (is (true? true))))))
```

Comment:
Yeah, I'm definitely good with deferring any additional changes, just trying to make sure I understand the current semantics.  If you're confident that, after the final unregister, the lock will be granted and there will be a poison pill... agh probably faster to discuss this in IRL